demos/coreHTTP/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,965 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
+ */
+
+/*
+ * Demo for showing use of the HTTP API using a server-authenticated network
+ * connection.
+ *
+ * This example, using a pre-signed URL, resolves a S3 domain, establishes a TCP
+ * connection, validates the server's certificate using the root CA certificate
+ * defined in the config header, then finally performs a TLS handshake with the
+ * HTTP server so that all communication is encrypted.
+ *
+ * Afterwards, an independent HTTP task is started, to read requests from the
+ * request queue and execute them using the HTTP Client library API, and write
+ * the corresponding responses to the response queue. The main thread sends
+ * requests to the request queue, which are used to download the S3 file in
+ * chunks using range requests. While doing so, the main thread reads responses
+ * from the response queue continuously until the entire file is received. If
+ * any request fails, an error code is returned.
+ *
+ * @Note: This demo requires user-generated pre-signed URLs to be pasted into
+ * demo_config.h. Please use the provided script "presigned_urls_gen.py"
+ * (located in Http_Demo_Helpers) to generate these URLs. For detailed
+ * instructions, see the accompanied README.md.
+ */
+
+/**
+ * @file http_demo_s3_download_multithreaded.c
+ * @brief Demonstrates usage of the HTTP library.
+ */
+
+/* Standard includes. */
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+#include "queue.h"
+
+/* Include demo config as the first non-system header. */
+#include "http_demo_s3_download_multithreaded_config.h"
+
+/* Include common demo header. */
+#include "aws_demo.h"
+
+/* Transport interface implementation include header for TLS. */
+#include "transport_secure_sockets.h"
+
+/* Common HTTP demo utilities. */
+#include "http_demo_utils.h"
+
+/* HTTP API header. */
+#include "core_http_client.h"
+
+/*------------- Demo configurations -------------------------*/
+
+/* Check that the TLS port of the server is defined. */
+#ifndef democonfigHTTPS_PORT
+    #error "Please define a democonfigHTTPS_PORT."
+#endif
+
+/* Check that the root CA certificate is defined. */
+#ifndef democonfigROOT_CA_PEM
+    #error "Please define a democonfigROOT_CA_PEM."
```

Comment:
Please also point to the file in which these MACROs are recommended to be defined.